### PR TITLE
Add support for Path from pathlib.

### PR DESCRIPTION
**Pull request by Hinidu**
Adding support for `Path` in `file_dep` is quite easy. But clearly this is only the part of the problem: if someone (like me) uses `Path` then he want to use them all other the place. It would be quite strange to pass them to `file_dep` as is but convert to `str` before passing in `targets` for example. I think at least `targets` should support them too and perhaps it would be not very bad idea to support them in `CmdAction`.
@schettino72 what do you think about it?

### CODE REVIEW

**Comment by schettino72** [comment]
Yes. Support it everywhere... 

I was even thinking about actually storing the PosixPath instead of converting it to string.
I would need to try implementing the code to see exactly what would be the consequences of that change...

**Comment by Hinidu** [comment]
> I was even thinking about actually storing the PosixPath instead of converting it to string.
I would need to try implementing the code to see exactly what would be the consequences of that change...

What's the benefit? I can imagine only two implementations:
1) Store in `file_dep` both `str`s and `Path`s and don't forget to cast items in `file_dep` when needed - cumbersome and error-prone.
2) Use `Path`s internally and convert `str`s provided by the user to them - pointless as I can see because I think `doit` doesn't use many operations on the file paths besides path comparison, though rich set of handy operations is the main feature of `pathlib` in my point of view.

But maybe I miss something important.

**Comment by schettino72** [comment]
> What's the benefit?

I was thinking about supporting passing a `glob` as a `file_dep` so the glob could be lazily evaluated. 
But the glob result is not even an instance of PosixPath, it is just a generator. 
So not really much relation to what you are doing...

Another reason would be to have a more robust file comparison. See #109, but I guess that is not worth the problem.

Conclusion: you are right :smile: , just convert to string.

**Comment by schettino72** [comment]
- You need to add a dependency to `pathlib` on `setup.py` because doit still supports python3.3. 

- Note some test failures on windows, check appveyor

**Comment by Hinidu** [comment]
> I was thinking about supporting passing a glob as a file_dep so the glob could be lazily evaluated. 
But the glob result is not even an instance of PosixPath, it is just a generator.
So not really much relation to what you are doing...

It also can be a good thing to do, but I agree that not in this PR.

> Another reason would be to have a more robust file comparison. See #109, but I guess that is not worth the problem.

Hmm, I didn't think about `some_file` and `./some_file`. I thought only about case-insensitive file systems.
Now it looks more benefitial. I will try to investigate the converting of all user strings to `Path`s - maybe it can be done as the next step after accepting `Path`s from the user.

> Conclusion: you are right :smile: , just convert to string.

Yeah, I think it would be a good first step :-)

> You need to add a dependency to pathlib on setup.py because doit still supports python3.3.
Note some test failures on windows, check appveyor

Ok.

**Comment by Hinidu** [comment]
> Note some test failures on windows, check appveyor

It looks like these failed tests didn't run before this PR on Windows. I assumed it by the number of tests:
- PR: `============== 4 failed, 702 passed, 15 skipped in 13.44 seconds ==============`
- master: `=================== 699 passed, 21 skipped in 13.94 seconds ===================`

I have added only one test.
These failed tests have a decorator: `@pytest.mark.skipif("os.system('strace -V') != 0")`
Do you have any ideas why `strace -V` could change its behaviour on AppVeyor?

**Comment by Hinidu** [comment]
@schettino72 I've commited support for `Path` in `targets` and `CmdAction`. Waiting for your comments on implementation. Could you point to the specific parts of the documentation which I should update to reflect these changes?

**Comment by Hinidu** [comment]
Perhaps it would be reasonable to do type checking for `targets` and `CmdAction` similar to existing for `file_dep` to provide more meaningful error message to the user.
Current approach will convert to `str` anything user will provide. I can imagine that some users want from time to time put some `int`s in `CmdAction`, but I think that other types in these places much rather would be a mistake.
What do you think about it?

**Comment by schettino72** [comment]
> Do you have any ideas why strace -V could change its behaviour on AppVeyor?

Sorry. No idea.

> Could you point to the specific parts of the documentation which I should update to reflect these changes?

When it introduces file_dep add an example using pathlib. Then in other places (targets, ...) just mention that it also supports pathlib too.

> Perhaps it would be reasonable to do type checking for targets and CmdAction similar to existing for file_dep to provide more meaningful error message to the user.

go ahead :smile: 

Sorry I didnt review the code yet... and probably wont do that next week also :sweat:

**Comment by schettino72** [comment]
> Do you have any ideas why strace -V could change its behaviour on AppVeyor?

I re-run the tests in an old commit just to double check.
Now AppVeyor provides a strace (from cygwin).
https://ci.appveyor.com/project/schettino72/doit/build/1.0.97/job/qlx1air685fiaxvs#L223

I dont want to debug that... lets just disable the test if strace not available OR it is on windows.
If you dont want to do this change yourself, it is fine... but please create another issue to track this. thanks.

**Comment by schettino72** [comment]
thanks. one more thing. add your name in the AUTHORS file :grinning:

**Comment by schettino72** [comment]
and an entry on CHANGES file

File: tests/test_task.py [hunk 1]
```diff
@@ -150,7 +151,12 @@ def test_dependencyStringIsFile(self):
         my_task = task.Task("Task X", ["taskcmd"], file_dep=["123","456"])
         assert set(["123","456"]) == my_task.file_dep
 
-    def test_file_dep_must_be_string(self):
+    def test_file_dep_can_be_path(self):
+        my_task = task.Task("Task X", ["taskcmd"],
+                            file_dep=["123", Path("456"), PurePath("789")])
+        assert {"123","456", "789"} == my_task.file_dep
+
+    def test_file_dep_must_be_string_or_path(self):
```
Comment (schettino72) [inline tests/test_task.py]: test_file_dep_path()

File: tests/test_task.py [hunk 2]
```diff
@@ -150,7 +151,12 @@ def test_dependencyStringIsFile(self):
         my_task = task.Task("Task X", ["taskcmd"], file_dep=["123","456"])
         assert set(["123","456"]) == my_task.file_dep
 
-    def test_file_dep_must_be_string(self):
+    def test_file_dep_can_be_path(self):
```
Comment (schettino72) [inline tests/test_task.py]: test_file_dep_path

File: tests/test_action.py [hunk 3]
```diff
@@ -222,6 +223,12 @@ def test_string_list_cant_be_expanded(self):
         my_action = action.CmdAction(cmd, task)
         assert cmd == my_action.expand_action()
 
+    def test_list_can_contain_path(self):
+        cmd = ["python", PurePath(TEST_PATH), Path("myecho.py")]
+        task = FakeTask([],[],[], {})
```
Comment (schettino72) [inline tests/test_action.py]: needs a whitespace after every comman

File: doit/action.py [hunk 4]
```diff
@@ -214,15 +214,17 @@ def execute(self, out=None, err=None):
 
 
     def expand_action(self):
-        """expand action string using task meta informations
-        @returns (string) - expanded string after substitution
+        """Expand action using task meta informations if action is a string.
+        Convert `Path` elements to `str` if action is a list.
```
Comment (schettino72) [inline doit/action.py]: better say exactly what you are doing (converting all elements to string), nothing specific about Path.

**Comment by schettino72** [comment]
looks good. thanks. just need docs now.

File: tests/test_action.py (hunk 3 above)
Comment (Hinidu) [inline tests/test_action.py]: Fixed

File: tests/test_task.py (hunk 2 above)
Comment (Hinidu) [inline tests/test_task.py]: Fixed

File: tests/test_task.py (hunk 1 above)
Comment (Hinidu) [inline tests/test_task.py]: Fixed

File: doit/action.py (hunk 4 above)
Comment (Hinidu) [inline doit/action.py]: I will add validation and raising `InvalidTask` if the user provides something that is nor `str` nor `Path`. I mentioned it earlier: https://github.com/pydoit/doit/pull/114#issuecomment-157982283

**Comment by Hinidu** [comment]
> thanks. one more thing. add your name in the AUTHORS file :grinning:

I'm already there :wink:

I'll add info to `CHANGES` and docs.

**Comment by Hinidu** [comment]
@schettino72 thank you for review.
I've pushed requested changes, type checking in `targets` and `CmdAction` and docs update.
PR looks ready to me now.

**Comment by schettino72** [comment]
thanks. squashed and merged.

There was a typo in the docs, already fixed. 
Next time you can check for typos using `doit spell`.

File: doc/tutorial/checker.py [hunk 5]
```diff
@@ -1,3 +1,10 @@
+from pathlib import Path
+
 def task_checker():
     return {'actions': ["pyflakes sample.py"],
             'file_dep': ["sample.py"]}
+
+def task_checker_pathlib():
```
Comment (schettino72) [inline doc/tutorial/checker.py]: actually I would prefer to have a separate example... to keep examples really simple.
But I merged anyway.

**Comment by Hinidu** [comment]
@schettino72 thank you for merging!
By the way this PR can help partially solve the problem from #109 if user will provide `Path` instances instead of strings. `str(Path('./afile')) == str(Path('afile')) == 'afile'`. But it doesn't resolve symbolic links and `..` though.

File: doc/tutorial/checker.py (hunk 5 above)
Comment (Hinidu) [inline doc/tutorial/checker.py]: I agree - it is a good thing to concentrate reader on one feature at once. I can remove this example here and add a new section like `pathlib support` with example which uses `pathlib` both in `file_dep`, `targets` and `CmdAction`.

Comment (schettino72) [inline doc/tutorial/checker.py]: PR with whatever improvements are always welcome.

**Comment by schettino72** [comment]
> By the way this PR can help partially solve the problem from #109 if user will provide Path instances instead of strings. str(Path('./afile')) == str(Path('afile')) == 'afile'. But it doesn't resolve symbolic links and .. though.

To tell you the truth I dont care about this feature. I am worried it could have a negative effect on performance. I dont know how I would implement that. If you to give a try go ahead :)

**Pull request closed without merging**